### ISSUE TITLE
Broken dmtcp_checkpoint() API

### ISSUE THREAD

**Issue by rohgarg** [dmtcp/dmtcp#215]
This issue was reported separately by @kwharrigan and @ankitcse07.

As @kwharrigan has rightly noted:

The commit 823d096 removed all references and the definition of the
function `dmtcp::userHookTrampoline_postCkpt()`.  The removal of
`userHookTrampoline_postCkpt` also removed calls that incremented the
static global variables `numRestarts` and `numCheckpoints`. Consequently,
the calls to `dmtcp_checkpoint()` would get stuck in the following while
loop in `dmtcpplugin.cpp`.

```C
  if(dmtcpRunCommand('c')){ //request checkpoint
    //and wait for the checkpoint
    while(oldNumRestarts==numRestarts && oldNumCheckpoints==numCheckpoints){
      //nanosleep should get interrupted by checkpointing with an EINTR
error
      //though there is a race to get to nanosleep() before the checkpoint
      struct timespec t = {1,0};
      nanosleep(&t, NULL);
      memfence();  //make sure the loop condition doesn't get optimized
    }
    rv = (oldNumRestarts==numRestarts ? DMTCP_AFTER_CHECKPOINT :
DMTCP_AFTER_RESTART);
  }
```

@kwharrigan has suggested the following patch to fix the problem:

```diff
diff --git a/include/dmtcp.h b/include/dmtcp.h
index 39ca249..1a4b6ac 100644
--- a/include/dmtcp.h
+++ b/include/dmtcp.h
@@ -417,4 +417,6 @@ DECLARE_TYPEOF_FNC(dmtcp_event_hook,_real_dmtcp_event_hook);
 /// Pointer to a "void foo();" function
 typedef void (*dmtcp_fnptr_t)(void);
 
+EXTERNC void increment_counters(int isRestart);
+
 #endif
diff --git a/src/dmtcpplugin.cpp b/src/dmtcpplugin.cpp
index 4c5167c..3b40fa1 100644
--- a/src/dmtcpplugin.cpp
+++ b/src/dmtcpplugin.cpp
@@ -500,3 +500,12 @@ EXTERNC int dmtcp_no_coordinator(void)
 {
   return CoordinatorAPI::noCoordinator();
 }
+
+EXTERNC void increment_counters(int isRestart)
+{
+  if (isRestart) {
+    numRestarts++;
+  } else {
+    numCheckpoints++;
+  }
+}
diff --git a/src/mtcpinterface.cpp b/src/mtcpinterface.cpp
index 282ba94..3ef2f6b 100644
--- a/src/mtcpinterface.cpp
+++ b/src/mtcpinterface.cpp
@@ -89,6 +89,8 @@ void dmtcp::callbackPostCheckpoint(int isRestart,
 
   DmtcpWorker::waitForStage4Resume(isRestart);
 
+  increment_counters(isRestart);
+
   WorkerState::setCurrentState( WorkerState::RUNNING );
 
   DmtcpWorker::informCoordinatorOfRUNNINGState();
```

@gc00 and @karya0 Could you please review the above patch?

**Comment by rohgarg** [comment]
Ping. :-)

**Comment by gc00** [comment]
@rohgarg, I've reviewed it now.  LGTM.

**Comment by rohgarg** [comment]
Closed via PR #217.
